src/ucs/datastruct/ptr_map.h

```diff
@@ -39,14 +40,131 @@ KHASH_TYPE(ucs_ptr_map_impl, ucs_ptr_map_key_t, void*);
 typedef khash_t(ucs_ptr_map_impl) ucs_ptr_hash_t;
 
 
+#define UCS_PTR_MAP_TYPE_COMMON \
+    uint64_t next_id; /**< Generator of unique keys per map. */ \
+    ucs_ptr_hash_t hash; /**< Hash table to store indirect key to pointer \
+                              associations. */
+
+
+#define UCS_PTR_MAP_TYPE_THREAD_SAFE_0
+#define UCS_PTR_MAP_TYPE_THREAD_SAFE_1 \
+    ucs_ptr_hash_t hash_safe; /**< Thread-safe hash table to store indirect \
+                                   key to pointer associations. */ \
+    ucs_spinlock_t lock; /**< Spin lock to synchronize access to hash_safe \
+                              hash table. */
+
+
 /**
  * Associative container key -> object pointer.
+ *
+ * @param [in]  _is_put_thread_safe  Defines whether UCS_PTR_MAP_PUT is thread
+ *                                   safe or not. If the operation is safe, then
+ *                                   it is safe to use the operation
+ *                                   concurrently with itself, UCS_PTR_MAP_GET
+ *                                   or UCS_PTR_MAP_DEL operations.
+ *
+ * @note Using UCS_PTR_MAP_GET and UCS_PTR_MAP_DEL with themselves and among
+ *       themselves is not thread safe regardless of @a _is_put_thread_safe.
+ */
+#define UCS_PTR_MAP_TYPE(_name, _is_put_thread_safe) \
+    typedef struct ucs_ptr_map_##_name { \
+        UCS_PTR_MAP_TYPE_COMMON; \
+        UCS_PTR_MAP_TYPE_THREAD_SAFE_##_is_put_thread_safe; \
```

Comment:
minor: could do this without macro by
```
struct {
  ...
  struct {
    ucs_ptr_hash_t hash_safe;
    ucs_spinlock_t lock;
  } safe[_is_put_thread_safe];
}
```

Reply:
\+ static assert for safe/unsafe cases to avoid safe[5] case